Loonyland Halloween Hill: implement new game

AutomaticFrenzy opened the pull request:
## What is this fixing or adding?

Adding a new game, Loonyland Halloween Hill

## How was this tested?

Ran 4 multiworld games with friends, 2 solo games.
Generated 25 successful 10 player loonyland multiworlds, with weighted options allowing for all combinations.

## If this makes graphical changes, please attach screenshots.

I haven't added it to the launcher yet, but I got permission from the original game creator to use the icon.

AutomaticFrenzy committed:
Base init and folder

AutomaticFrenzy committed:
Add base docs

AutomaticFrenzy committed:
add base files

AutomaticFrenzy committed:
base items commit

AutomaticFrenzy committed:
forgot to save

AutomaticFrenzy committed:
Ids for Items
locations
regions

AutomaticFrenzy committed:
location ids and categories
all da rules

AutomaticFrenzy committed:
python syntax

AutomaticFrenzy committed:
add logical regions
formatting and cleanup

AutomaticFrenzy committed:
base options

AutomaticFrenzy committed:
naming, typos

AutomaticFrenzy committed:
base init.py

AutomaticFrenzy committed:
renaming, lots of work in init.py

AutomaticFrenzy committed:
naming locations to avoid duplicates
remove bad copied code from regions
Support location_name_to_id and item_name_to_id in init.py

AutomaticFrenzy committed:
Entrance Information
Locations include region in their name
regions implemented
Rules cleanup
Working on regions and entrances in init

AutomaticFrenzy committed:
remove entrance rules, put it in entrance table
add rules to entrance table
new logical regions

AutomaticFrenzy committed:
start adding exits to regions

AutomaticFrenzy committed:
entrance adding to region code
capitalization issues

AutomaticFrenzy committed:
Name fixing

AutomaticFrenzy committed:
hey its generating

AutomaticFrenzy committed:
Should be working
Name typos
Win con event added

AutomaticFrenzy committed:
fix tests

AutomaticFrenzy committed:
move logic to regions and clean up rules file
put zombie stomp into zombiton

AutomaticFrenzy committed:
Renaming zones, starting on extra subregions for ER

AutomaticFrenzy committed:
auto formatting pass

AutomaticFrenzy committed:
More entrance/region work

AutomaticFrenzy committed:
1.0 working

AutomaticFrenzy committed:
switch to auto generated data objects

AutomaticFrenzy committed:
add difficulty as an option

AutomaticFrenzy committed:
autogenerated data

AutomaticFrenzy committed:
Send deathlink info

AutomaticFrenzy committed:
Simplify entrance data

AutomaticFrenzy committed:
Cleanup
data in one .py
update setup and main game docs
set entrance rules in set_rules

AutomaticFrenzy committed:
Merge branch 'ArchipelagoMW:main' into ArchipelagoLoonyland

AutomaticFrenzy committed:
add base id to location

AutomaticFrenzy committed:
Renaming, prettifying, etc
Handle 39 badges and all special weapons rules

AutomaticFrenzy committed:
ruff format

AutomaticFrenzy committed:
junk filler
access items
more data

AutomaticFrenzy committed:
Doll support

AutomaticFrenzy committed:
Handling more options

AutomaticFrenzy committed:
power level, monster doll stuff, options handling

AutomaticFrenzy committed:
more logic to items and locations
new multiplesaves multisave option
entrances and regions have flags now

AutomaticFrenzy committed:
add some locs as postgame

AutomaticFrenzy committed:
add witch reward to item pool
badge WinCondition option
badges required option
op cheats option
events for badge earn locs, so I can check how many you have in logic gen

AutomaticFrenzy committed:
data updates

AutomaticFrenzy committed:
logic updates for long tag on two badges and ghost potion in haunted tower

AutomaticFrenzy committed:
logic updates
testing

AutomaticFrenzy committed:
remove rerunning default test, run_default_tests takes care of it

AutomaticFrenzy committed:
data updates
explicit options imports for ruff

AutomaticFrenzy committed:
guide updates

AutomaticFrenzy committed:
Merge branch 'main' of https://github.com/ArchipelagoMW/Archipelago into ArchipelagoLoonyland

AutomaticFrenzy committed:
relative imports

AutomaticFrenzy committed:
add VANILLA_POSTGAME for badges that are postgame locs in BADGES_VANILLA

ScipioWright commented on worlds/loonyland/__init__.py:
```diff
@@ -0,0 +1,156 @@
+
+
+class LoonylandWorld(World):
+    """The greatest game of all time."""
```
Definitely change this

ScipioWright commented on worlds/loonyland/__init__.py:
```diff
@@ -0,0 +1,156 @@
+        for loc_name, loc_data in loonyland_location_table.items():
+            if not loc_data.can_create(self.options):
+                continue
+            region = self.multiworld.get_region(loc_data.region, self.player)
```
```suggestion
            region = self.get_region(loc_data.region)
```
Just a shorter version of it

ScipioWright commented on worlds/loonyland/__init__.py:
```diff
@@ -0,0 +1,156 @@
+
+    def set_rules(self):
+        # Completion condition.
+        # self.multiworld.completion_condition[self.player] = lambda state: state.has("Victory", self.player)
```
```suggestion
```

ScipioWright reviewed (commented):
Just a few tiny things, don't have much energy to review atm

AutomaticFrenzy committed:
better game description, removing comment, simplifying region line

Mysteryem commented on worlds/loonyland/Data/game_data.py:
```diff
@@ -0,0 +1,1308 @@
+        ),
+        LLEntrance("Rocky Cliffs", "Abandoned Mines Entrance", True),
+        LLEntrance("Rocky Cliffs", "The Shrine Of Bombulus", True),
+        LLEntrance("Rocky Cliffs", "A Gloomy Cavern Entrance", True, lambda state: have_light_source(state, world)),
```
`have_light_source` checks for being able to reach the `Swamp Gas Cavern` region, so every entrance that uses `have_light_source` in its access rule requires the `Swamp Gas Cavern` region to be registered as an indirect condition:
https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/world%20api.md#an-important-note-on-entrance-access-rules

AutomaticFrenzy replied:
That location is where you light a stick to make it a torch. I added a torch item and forced it to be at that location now, so I can just check for the item.

AutomaticFrenzy committed:
rename a cool filler item to its effect, "Max life and gems"
add torch item, place it at the curse the darkness location
make torch item requirement for light source instead of reaching swamp gas cavern

nicholassaylor commented on worlds/loonyland/__init__.py:
```diff
@@ -0,0 +1,164 @@
+
+class LoonylandWorld(World):
+    """
+    Loonyland: Halloween Hill is an This is an action-adventure game,
```
```suggestion
    Loonyland: Halloween Hill is an action-adventure game,
```

AutomaticFrenzy committed:
Update worlds/loonyland/__init__.py

Co-authored-by: Nicholas Saylor <79181893+nicholassaylor@users.noreply.github.com>

beauxq commented on worlds/loonyland/rules.py:
```diff
@@ -0,0 +1,130 @@
+
+from BaseClasses import CollectionState
+
+if TYPE_CHECKING:
+    from worlds.loonyland import LoonylandWorld
```
It looks like every use of `LoonylandWorld` in this file is just getting the `player` attribute.

Maybe `player` could be passed to the functions instead, and then this `if TYPE_CHECKING:` import wouldn't be needed.
But maybe there are plans to use it for more in the future... I don't know.

AutomaticFrenzy replied:
[9381051](https://github.com/ArchipelagoMW/Archipelago/pull/4427/commits/938105182c1f8ffa94c8a9b8c8c88fe19e530068)

beauxq commented on worlds/loonyland/ruff.toml:
```diff
@@ -0,0 +1,9 @@
+select = ["C", "E", "F", "R", "W", "I", "N", "Q", "UP", "RUF", "ISC", "T20"]
+ignore = ["C9", "RUF012", "RUF100"]
+
+[lint.per-file-ignores]
+# The way options definitions work right now, world devs are forced to break line length requirements.
+"options.py" = ["E501"]
```
If I'm understanding this right, I think this is the solution to this, and the recommended way for new worlds:
https://github.com/ArchipelagoMW/Archipelago/blob/main/docs/options%20api.md#option-documentation

```python
class LoonylandWebWorld(WebWorld):
    # Render all this world's options as rich text.
    rich_text_options_doc = True
```

AutomaticFrenzy replied:
[e5c27e7](https://github.com/ArchipelagoMW/Archipelago/pull/4427/commits/e5c27e768013bb08268d1f9c4e4ecb6e1b5fdcf0)

beauxq commented on worlds/loonyland/entrances.py:
```diff
@@ -0,0 +1,27 @@
+    is_real_loading_zone: bool
+    # rule: typing.Callable[[player, state], bool]
+    rule: CollectionRule = lambda state: True
+    flags: list[str] = []
```
This applies to `LLEntrance` and `LLItem` and `LLLocation` and `LLRegion`:

My first concern with this is the mutable default value, which is dangerous.
So I thought of suggesting a change to `Sequence[str]` and `()` default value.

But then I looked at how they are used, and I think an `enum.Flag` would make more sense.

AutomaticFrenzy replied:
[a1d9294](https://github.com/ArchipelagoMW/Archipelago/pull/4427/commits/a1d92942376f59977010bcd7ceca0ce6918acdb5)

AutomaticFrenzy committed:
Options page is now using rich text
Fix setup guide not being linked correctly
Change toml to affect options.py now

AutomaticFrenzy committed:
Change flags from being a list of strings into a enum.flag

AutomaticFrenzy committed:
Merge branch 'main' of https://github.com/ArchipelagoMW/Archipelago into ArchipelagoLoonyland

AutomaticFrenzy committed:
remove "world" references from the rules, just pass in player

beauxq commented on worlds/loonyland/options.py:
```diff
@@ -0,0 +1,126 @@
+
+    None: Badges aren't in logic"""
+
+    display_name = "Badges"
+    option_none = 0
+    option_vanilla = 1
+    option_full = 2
+    default = 2
```
The documentation has something called "Reasonable"
But that's missing from the options here.

AutomaticFrenzy replied:
Good catch, that option is Long Checks now, so I removed it from the doc string

[8214b9c](https://github.com/ArchipelagoMW/Archipelago/pull/4427/commits/8214b9c8b93f9a3d50faf56678fd8f5531678d0a)

AutomaticFrenzy committed:
formatting updates for options descriptions

AutomaticFrenzy commented on worlds/loonyland/options.py:
```diff
@@ -0,0 +1,126 @@
+
+    Vanilla: Badges aren't randomized, but are in logic
+
+    None: Badges aren't in logic"""
```
Went with the formatting that Accessibility uses, bold for the options text and newlines.

[8214b9c](https://github.com/ArchipelagoMW/Archipelago/pull/4427/commits/8214b9c8b93f9a3d50faf56678fd8f5531678d0a)

AutomaticFrenzy replied:
![image](https://github.com/user-attachments/assets/f1c5fefa-79eb-400b-a0c3-c981328282fe)

AutomaticFrenzy committed:
Raise OptionError if the yaml has more badges required than in logic for wincon

AutomaticFrenzy committed:
Logic updates
reclassified some cheats as non-prog
No longer require ghost potion for sneakin'
add skull key req for right vine in under the ravine
add power level requirement for happy stick witch quest fight
add power level req for a hidey hole
Dont require frog luring for swamp gas cavern bomb rocks + combo bombo

qwint commented on worlds/loonyland/test/__init__.py:
FYI new guidance any test base should be in a `/test/bases.py` file and the `/test/__init__.py` should be an empty file for test discovery

qwint commented on worlds/loonyland/__init__.py:
```diff
@@ -0,0 +1,166 @@
+                    new_item = self.create_item(name)
+                    item_pool.append(new_item)
+
+        junk_len = len(self.multiworld.get_unfilled_locations(self.player)) - len(item_pool) - 1
```
i believe this ` - 1` is for the locked item you place in set_rules correct? should probably document that (or just move the item placement earlier to create_regions so the offset isn't needed)

qwint commented on worlds/loonyland/__init__.py:
```diff
@@ -0,0 +1,166 @@
+        "badges": {name for name, data in loonyland_location_table.items() if data.category == LLLocCat.BADGE},
+    }
+
+    def create_junk(self) -> LoonylandItem:
+        return LoonylandItem("Max Life and Gems", ItemClassification.filler, loonyland_base_id + 3000, self.player)
```
FYI the expected way to create excess filler is define a get_filler_item_name <https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/AutoWorld.py#L485>
and then use that in your create_item (or use the helper that does that automatically <https://github.com/ArchipelagoMW/Archipelago/blob/main/worlds/AutoWorld.py#L544>)

AutomaticFrenzy replied:
[340d34f](https://github.com/ArchipelagoMW/Archipelago/pull/4427/commits/340d34f4a6b37607edc00dae598c73f2d6e24e08)

AutomaticFrenzy committed:
Merge branch 'ArchipelagoMW:main' into ArchipelagoLoonyland

AutomaticFrenzy committed:
Merge branch 'ArchipelagoLoonyland' of https://github.com/AutomaticFrenzy/Archipelago into ArchipelagoLoonyland

AutomaticFrenzy committed:
Move tests to bases.py, empty __init.py__

AutomaticFrenzy committed:
Fix rules logic about gems

AutomaticFrenzy committed:
Test was erroring due to game_data.py in its own folder

AutomaticFrenzy committed:
Documentation updates

AutomaticFrenzy committed:
More Doc updates

AutomaticFrenzy committed:
Remove unecessary multiworld arguments

AutomaticFrenzy committed:
Move win condition logic earlier to simplify item creation

AutomaticFrenzy committed:
Reorganize junk creation to be the expected way

AutomaticFrenzy committed:
renamed logic to other words to avoid confusion with access requirements

AutomaticFrenzy commented on worlds/loonyland/locations.py:
```diff
@@ -0,0 +1,68 @@
+            return False
+        return True
+
+    def in_logic(self, options: LoonylandOptions) -> bool:
```
It controls whether some items/locations are included in the pool or not. Renamed in https://github.com/ArchipelagoMW/Archipelago/pull/4427/commits/12bbb4608eff95c9830de2e0ad390fc0060ed94c

AutomaticFrenzy committed:
simplify power level with a collect override

AutomaticFrenzy committed:
simplify slot data, client updated to expect new values

AutomaticFrenzy commented on worlds/loonyland/__init__.py:
```diff
@@ -0,0 +1,166 @@
+        # entrance rules
+        set_entrance_rules(self.multiworld, self)
+
+    def fill_slot_data(self):
```
https://github.com/ArchipelagoMW/Archipelago/pull/4427/commits/21527ece46a2e58d355260043fbb834ecdd2945d

qwint reviewed (approved):
My previous comments were addressed

didn't do a full re-review, I trust my previous review (and had marked all files as viewed so it was complete)
did not do any test runs/gens/etc. purely a code review

AutomaticFrenzy committed:
Commit entrance rando changes, still WIP

AutomaticFrenzy committed:
Merge remote-tracking branch 'remotes/Upstream/main' into ArchipelagoLoonyland

AutomaticFrenzy committed:
Logic update for daisy quest

AutomaticFrenzy committed:
add manifest
